### PR TITLE
constexpr implementation of KMP Algorithm

### PR DESCRIPTION
**Pull request by serge-sans-paille**
https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
@jaydee-io any though on that one? Your previous review were very valuable!

**Comment by jaydee-io** [comment]
Personally, my preference goes to [Boyer-Moore](https://en.wikipedia.org/wiki/Boyer–Moore_string_search_algorithm#Performance) or [Boyer-Moore-Horspool](https://en.wikipedia.org/wiki/Boyer–Moore–Horspool_algorithm) algorithm (those ones that have been added to C++17 searcher `std::boyer_moore_searcher` and `std::boyer_moore_horspool_searcher`) ;-)
But I'll have a look at your KMP.

File: tests/Makefile [hunk 1]
```diff
@@ -45,3 +45,7 @@ test_unordered_str_set.o: test_unordered_str_set.cpp \
   ../include/frozen/bits/basic_types.h ../include/frozen/bits/elsa.h \
   ../include/frozen/string.h \
   catch.hpp
+test_str.o: test_str.cpp \
```
Comment (jaydee-io) [inline tests/Makefile]: It seems that your commit is missing`test_str.cpp`

File: include/frozen/string.h [hunk 2]
```diff
@@ -70,12 +71,91 @@ template <> struct elsa<string> {
...
+    constexpr char operator[](std::size_t i) const {
+      return (&std::get<0>(data_))[i];
+    }
+    constexpr std::ptrdiff_t step(std::size_t i) const {
+      return cache_[i];
+    }
+};
+
+template<std::size_t N>
+std::size_t strstr(std::string const& haystack, string_needle<N> const& needle) {
```
Comment (jaydee-io) [inline include/frozen/string.h]: Maybe a comment mentioning that this function is using the KMP search algorithm may be useful when we'll reading back this code in several years...

File: include/frozen/string.h [hunk 3]
```diff
@@ -70,12 +71,91 @@ template <> struct elsa<string> {
   }
 };
 
+template<std::size_t size>
+class string_needle {
+  bits::cvector<std::ptrdiff_t, size> cache_;
```
[the remark this answers is not in the document]
Comment (jaydee-io) [reply include/frozen/string.h]: Also, `cache_` should be `size + 1` long. See below for the explanation.

File: include/frozen/string.h [hunk 4]
```diff
@@ -70,12 +71,91 @@ template <> struct elsa<string> {
...
 namespace string_literals {
 
 constexpr string operator"" _s(const char *data, std::size_t size) {
   return {data, size};
 }
+
+
 } // namespace string_literals
+
+#define FROZEN_NEEDLE(s) frozen::string_needle<sizeof(s) - 1>{s}
```
Comment (jaydee-io) [inline include/frozen/string.h]: Couldn't we get rid of the macro with adding this constructor instead ?

```
constexpr string_needle(char const (&data)[size - 1])
```

or

```
template<std::size_t strSize = size - 1>
constexpr string_needle(char const (&data)[strSize])
```

**Comment by jaydee-io** [comment]
Oh... and usually, I prefer `x++` over `x += 1` :-)

File: include/frozen/string.h [hunk 5]
```diff
@@ -70,12 +71,91 @@ template <> struct elsa<string> {
...
+        cnd += 1;
+      }
+      else {
+        cache[pos] = cnd;
+        cnd = cache[cnd];
+        while(cnd >= 0 && Needle[pos] != Needle[cnd])
+          cnd = cache[cnd];
+        cnd += 1;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (serge-sans-paille) [reply include/frozen/string.h]: I think it's not necessay; It just supress the check for the trailing zero and repalce it by a size check.

File: include/frozen/string.h (hunk 2 above)
Comment (serge-sans-paille) [reply include/frozen/string.h]: Yes! I'll a wikipedia ref and provide an interface similar to the std::search you've been pointing out.

**Comment by serge-sans-paille** [comment]
@jaydee-io thanks for your review; If that's OK with you I'll ad a word about your contribution in the thanks!

**Comment by jaydee-io** [comment]
@serge-sans-paille It's always a pleasure if I could help. No problem to mention my contribution.

**Pull request merged**